packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -1,65 +1,190 @@
 import { parseVatSlot } from './parseVatSlots.js';
 
+/*
+  Theory of Operation (vref logical objects)
+
+  Liveslots coordinates access to "logical objects", sometimes known
+  as "vref objects" because each one is named with a distinct vref
+  identifier (o+1, o-2, o+d3/4, etc). The SwingSet kernel coordinates
+  access to the objects between vats, using a kref identifier (ko56),
+  but vats never see krefs.
+
+  When vat code (written in JavaScript) needs to interact with a
+  logical object, it needs some JS `Object` to pass in arguments, or
+  to invoke methods upon, or to receive in the arguments of a method
+  call. We use Presence objects for imported vrefs, Remotable/heap
+  objects for locally-created (potentially-exported) ephemeral
+  objects, and Representatives for virtual/durable objects initially
+  created by calling the Kind's maker function. Liveslots is
+  responsible for recognizing these `Object` objects when they appear
+  at its doorstep (e.g. in a remote method send, or when assigned to
+  the `state` of a virtual object), and knowing which vref is being
+  referenced. Liveslots is also the one to create all Presences and
+  Representatives (vat code creates the Remotable/heap objects, by
+  calling `Far()`).
+
+  For garbage-collection purposes, liveslots tracks the "reachable"
+  and "recognizable" status for these logical objects.  A logical
+  object is either VREF-REACHABLE, VREF-RECOGNIZABLE, or nothing. The
+  logical objects form a two-color directed graph, or you may think of
+  it as a graph and an overlaying subgraph. The first color contains
+  "weak" or "recognizable" edges: a weakly-keyed collection can
+  recognize its keys, but does not hold them strongly, so it does not
+  provide "reachable" access. The second/overlay color represents
+  "strong" or "reachable" edges, where access to the first object
+  enables strong access to the second.
```

Comment:
Yeah, I guess if I'm going to try to interpret this as a multi-colored graph, I have to work out the details more completely. The weak-colored edge that a weak collection provides from key to value is itself held alive by the collection, which means there must be a strong-color edge from something in the REACHABLE set to the collection itself.

I think I'll remove the details and just say that there's a graph-based interpretation without actually trying to nail down what it is. I think the reader will understand the intent.

Reply:
```

  For garbage-collection purposes, liveslots tracks the "reachable"
  and "recognizable" status for these logical objects.  A logical
  object is either VREF-REACHABLE, VREF-RECOGNIZABLE, or
  nothing. Strong references from a VREF-REACHABLE object makes the
  target also VREF-REACHABLE. Weak references from a VREF-REACHABLE
  makes the target at least VREF-RECOGNIZABLE, although it might be
  VREF-REACHABLE if there is also a strong reference to it.

  Weak collections provide weak access to their keys, and strong
  access to their values. Strong collections provide strong access to
  both keys and values. Virtual/durable objects provide strong access
  to their `state`. The "reachable set" is the transitive closure of
  strong edges, given a set of anchors. There will also be a
  "recognizable set", with weak edges from some members of the
  reachable set.
```